lib/models/cql_calculator.js

```diff
@@ -0,0 +1,643 @@
+/**
+ * The CQL calculator. This calls the cql-execution framework and formats the results as neccesary for Bonnie
+ * and Cypress.
+ */
+
+const _ = require('lodash');
+const moment = require('moment');
+const CqmModels = require('cqm-models');
+const Cql = require('cqm-models').CQL;
+const Measure = require('cqm-models').Measure();
+
+module.exports = class CQLCalculator {
+  constructor() {
+    this.calculator = {};
+    this.calculatorLoaded = {};
+    this.resultsCache = {};
+    this.valueSetsByMongoidCached = {};
+  }
+
+  // Key for storing calculators on a measure
+  static calculationKey(measure) { return `${measure.id}`; }
+
+  // Key for storing results for a patient / population calculation
+  // we use the mongo ID for the patient portion
+  cacheKey(measure, patient) { return `${this.calculationKey(measure)}/${patient.id()}`; }
+
+  clearResult(measure, patient) {
+    return delete this.resultsCache[this.cacheKey(measure, patient)];
+  }
+
+  /**
+   * Generate a calculation result for a population / patient pair; this always returns a result immediately,
+   * but may return a blank result object if there was a problem. Currently we do not do CQL calculations in
+   * deferred manner like we did for QDM calcuations.
+   * @param {Measure} measure - The measure population to calculate on.
+   * @param {PatientSource} patientSource - The patientSource to run calculations on.
+   * @param {Hash} valueSetsByMongoid - all ValueSets relevant to the measure hashed by Mongo ID
+   * @param {Hash} options - contains options for measure calculation, particularly relevant is effective_date
+   */
+  calculate(measure, patientSource, valueSetsByMongoid, options) {
+    // We store both the calculation result and the calculation code based on keys derived from the arguments
+    let generatedELMJSON;
+
+    const resultsByPatient = {};
+
+    this.valueSetsByMongoid = valueSetsByMongoid;
+
+    patientSource.reset();
+
+    // try {
+    // Grab start and end of Measurement Period
+    let start;
+    let end;
+    // Override default measure_period with effective_date if available
+    if (options && options.effective_date != null) {
+      start = this.constructor.getConvertedTime(options.effective_date);
+      if (options.effective_date_end != null) {
+        end = this.constructor.getConvertedTime(options.effective_date_end);
+      } else {
+        end = this.constructor.getConvertedTimeEndOfYear(options.effective_date);
+      }
+    } else {
+      start = this.constructor.getConvertedTime(measure.get('measure_period').low.value);
+      end = this.constructor.getConvertedTime(measure.get('measure_period').high.value);
+    }
+    const startCql = Cql.DateTime.fromDate(start, 0); // No timezone offset for start
+    const endCql = Cql.DateTime.fromDate(end, 0); // No timezone offset for stop
+
```

Comment:
There is another "parameter" that was added in the Bonnie calculator. `executionDateTime = cql.DateTime.fromDate(new Date(), '0')`  This creates the execution DateTime that we pass into the engine. Essentially telling the engine what the current time is when the execution starts. There is more code that will need to be updated further down in this file.